torchao/quantization/quantize_/workflows/float8/float8_tensor.py

```diff
@@ -0,0 +1,609 @@
+        act_quant_kwargs: Optional[QuantizeTensorToFloat8Kwargs] = None,
+    ):
+        lp_dtype = quant_kwargs.dtype
+        granularity = quant_kwargs.granularity
+        mm_config = quant_kwargs.mm_config
+        block_size = get_block_size(hp_tensor.shape, granularity)
+        block_size = list(block_size)
+
+        # TODO: https://github.com/pytorch/ao/issues/2511
+        # NOTE: fbgemm quant primitives like `torch.ops.fbgemm.quantize_fp8_per_row`
+        # can't pass some of the numerics tests, so we use torchao ones
+        scale = _choose_qparams_affine_float8(
+            hp_tensor, float8_dtype=lp_dtype, block_size=block_size
+        )
+        data = _quantize_affine_float8(hp_tensor, scale, lp_dtype)
+        hp_dtype = hp_tensor.dtype
+        return Float8Tensor(
+            data,
+            scale,
+            block_size=block_size,
+            mm_config=mm_config,
+            act_quant_kwargs=act_quant_kwargs,
+            dtype=hp_dtype,
+        )
+
+
+implements = Float8Tensor.implements
+
+
+@implements([torch.nn.functional.linear, aten.linear.default])
+def _(func, types, args, kwargs):
+    input_tensor, weight_tensor, bias = (
+        args[0],
+        args[1],
+        args[2] if len(args) > 2 else None,
+    )
+    if not isinstance(weight_tensor, Float8Tensor):
+        raise NotImplementedError("Only support handling Float8 weight")
+
+    # NOTE: we are using torchao quant primitive ops instead of fbgemm ones due to various numerical
+    # issues with fbgemm quant primitive ops
+    # for example fbgemm per tensor quant op is not as accurate as the quant primitives in torchao and won't
+    # pass the accuracy check in
+    # TestAffineQuantizedFloat8Compile.test_float8_tensor_slicing_functional_correctness
+    # https://github.com/pytorch/ao/blob/20d45036603c96873b8a8a8391fdbf2a2771ab7e/test/dtypes/test_affine_quantized_float.py#L631
+    # a_data, a_scale = torch.ops.fbgemm.quantize_fp8_per_tensor(input_tensor)
+    act_quant_kwargs = weight_tensor.act_quant_kwargs
+    if act_quant_kwargs is not None:
+        input_tensor = _choose_quant_func_and_quantize_tensor(
+            input_tensor, act_quant_kwargs
+        )
+
+    if isinstance(input_tensor, Float8Tensor):
+        scaled_mm_config = weight_tensor.mm_config
+        assert scaled_mm_config is not None
+        out_shape = get_out_shape(input_tensor.shape, weight_tensor.shape)
+
+        # Extract tensor data and scales
+        inpt_data = input_tensor._data.reshape(-1, input_tensor._data.shape[-1])
+        w_data = weight_tensor._data
+        input_scale = input_tensor.scale
+        w_scale = weight_tensor.scale.contiguous()
+
+        # Handle rowwise scaling
+        if _is_rowwise_scaled(weight_tensor):
+            assert _is_rowwise_scaled(input_tensor), (
+                "Input tensor must be rowwise block size"
+            )
+            w_scale = w_scale.transpose(-1, -2)
```

Comment:
curious why we need the transpose

Reply:
this is from original float8 code, probably just align with the requirement of the scaled_mm kernel: https://github.com/pytorch/ao/blob/11ce634705af1972320a8488d3e2ca223dcb0372/torchao/dtypes/floatx/float8_layout.py#L412-L416, @drisspg should have more context

Reply:
yeah it so that we get a scale of shape (1, N) for the second input to the Matrix and we know that we have 1 scale per reduction dim and we are doing rowwise scaling